src/postgres-client/src/postgres_client.rs

```diff
@@ -0,0 +1,219 @@
+        let pool = Pool::builder(manager)
+            .max_size(config.knobs.connection_pool_max_size())
+            .post_create(Hook::async_fn(move |client, _| {
+                connections_created.inc();
+                Box::pin(async move {
+                    debug!("opened new consensus postgres connection");
+                    client.batch_execute(
+                        "SET SESSION CHARACTERISTICS AS TRANSACTION ISOLATION LEVEL SERIALIZABLE",
+                    ).await.map_err(|e| HookError::Abort(HookErrorCause::Backend(e)))
+                })
+            }))
+            .pre_recycle(Hook::sync_fn(move |_client, conn_metrics| {
+                // proactively TTL connections to rebalance load to Postgres/CRDB. this helps
+                // fix skew when downstream DB operations (e.g. CRDB rolling restart) result
+                // in uneven load to each node, and works to reduce the # of connections
+                // maintained by the pool after bursty workloads.
+
+                // add a bias towards TTLing older connections first
+                if conn_metrics.age() < config.knobs.connection_pool_ttl() {
+                    return Ok(());
+                }
+
+                let last_ttl = last_ttl_connection.load(Ordering::SeqCst);
+                let now = (SYSTEM_TIME)();
+                let elapsed_since_last_ttl = Duration::from_millis(now.saturating_sub(last_ttl));
+
+                // stagger out reconnections to avoid stampeding the DB
+                if elapsed_since_last_ttl > config.knobs.connection_pool_ttl_stagger()
+                    && last_ttl_connection
+                        .compare_exchange_weak(last_ttl, now, Ordering::SeqCst, Ordering::SeqCst)
+                        .is_ok()
+                {
+                    ttl_reconnections.inc();
+                    return Err(HookError::Continue(Some(HookErrorCause::Message(
+                        "connection has been TTLed".to_string(),
+                    ))));
+                }
+
+                Ok(())
+            }))
+            .build()
+            .expect("postgres connection pool built with incorrect parameters");
+
+        Ok(PostgresClient {
+            pool,
+            metrics: config.metrics,
+        })
+    }
+
+    /// Gets connection from the pool or waits for one to become available.
+    pub async fn get_connection(&self) -> Result<Object, PoolError> {
+        let start = Instant::now();
+        let res = self.pool.get().await;
+        if let Err(PoolError::Backend(err)) = &res {
+            debug!("error establishing connection: {}", err);
+            self.metrics.connpool_connection_errors.inc();
+        }
+        self.metrics
+            .connpool_acquire_seconds
+            .inc_by(start.elapsed().as_secs_f64());
+        self.metrics.connpool_acquires.inc();
+        // note that getting the pool size here requires briefly locking the pool
```

Comment:
Heads up that you might end up conflicting with https://github.com/MaterializeInc/materialize/pull/21955 here

Reply:
@bkirwi do you want to go first or should I?

Reply:
Eh, go ahead... you've got the larger of the two changes.

Reply:
🙇‍♂️